clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -2357,11 +2355,9 @@ mlir::Value CIRGenFunction::buildLoadOfScalar(Address Addr, bool Volatile,
                                               QualType Ty, mlir::Location Loc,
                                               LValueBaseInfo BaseInfo,
                                               bool isNontemporal) {
-  if (!CGM.getCodeGenOpts().PreserveVec3Type) {
-    if (Ty->isVectorType()) {
-      llvm_unreachable("NYI");
-    }
-  }
+  if (!CGM.getCodeGenOpts().PreserveVec3Type && Ty->isVectorType() &&
+      Ty->castAs<clang::VectorType>()->getNumElements() == 3)
+    llvm_unreachable("NYI: Special treatment of 3-element vectors");
```

Comment:
Can you move this to after the atomic check and wrap it around `if (const auto *ClangVecTy = Ty->getAs<VectorType>()) {`? No problem if this comes in a later PR, but I'd prefer if the skeleton is a bit more similar. Similar for `buildStoreOfScalar` on the `if (const auto *ClangVecTy = Ty->getAs<VectorType>()) {` part.